mldsa/sign.c

```diff
@@ -862,3 +862,242 @@ int crypto_sign_open(uint8_t *m, size_t *mlen, const uint8_t *sm, size_t smlen,
+        return 0;
+      }
+      break;
+    case MLD_SHA3_224:
+      if (len == 28)
+      {
+        mld_memcpy(oid, sha3_224_oid, 11);
+        return 0;
+      }
+      break;
+    case MLD_SHA3_256:
+      if (len == 32)
+      {
+        mld_memcpy(oid, sha3_256_oid, 11);
+        return 0;
+      }
+      break;
+    case MLD_SHA3_384:
+      if (len == 48)
+      {
+        mld_memcpy(oid, sha3_384_oid, 11);
+        return 0;
+      }
+      break;
+    case MLD_SHA3_512:
+      if (len == 64)
+      {
+        mld_memcpy(oid, sha3_512_oid, 11);
+        return 0;
+      }
+      break;
+    case MLD_SHAKE_128:
+      if (len == 32)
+      {
+        mld_memcpy(oid, shake_128_oid, 11);
+        return 0;
+      }
+      break;
+    case MLD_SHAKE_256:
+      if (len == 64)
+      {
+        mld_memcpy(oid, shake_256_oid, 11);
+        return 0;
+      }
+      break;
+  }
+  return -1;
+}
+
+MLD_MUST_CHECK_RETURN_VALUE
+MLD_EXTERNAL_API
+int crypto_sign_signature_pre_hash_internal(uint8_t *sig, size_t *siglen,
+                                            const uint8_t *ph, size_t phlen,
+                                            const uint8_t *ctx, size_t ctxlen,
+                                            const uint8_t rnd[MLDSA_RNDBYTES],
+                                            const uint8_t *sk,
+                                            mld_hash_alg_t hashAlg)
+{
+  /* formatted message: 0x01 || ctxlen (1 byte) || ctx || oid || ph */
+  MLD_ALIGN uint8_t fmsg[2 + 255 + 11 + 64];
+  int result;
+
+  if (ctxlen > 255)
+  {
+    *siglen = 0;
+    return -1;
+  }
+  fmsg[0] = 1;
+  fmsg[1] = ctxlen;
+  if (ctx != NULL && ctxlen != 0)
+  {
+    mld_memcpy(fmsg + 2, ctx, ctxlen);
+  }
+  if (sha_oid(fmsg + 2 + ctxlen, hashAlg, phlen))
+  {
+    *siglen = 0;
+    return -1;
+  }
+  mld_memcpy(fmsg + 2 + ctxlen + 11, ph, phlen);
+
+  result = crypto_sign_signature_internal(
+      sig, siglen, fmsg, 2 + ctxlen + 11 + phlen, NULL, 0, rnd, sk, 0);
+  /* @[FIPS204, Section 3.6.3] Destruction of intermediate values. */
+  mld_zeroize(fmsg, sizeof(fmsg));
+  return result;
+}
+
+MLD_MUST_CHECK_RETURN_VALUE
+MLD_EXTERNAL_API
+int crypto_sign_verify_pre_hash_internal(const uint8_t *sig, size_t siglen,
+                                         const uint8_t *ph, size_t phlen,
+                                         const uint8_t *ctx, size_t ctxlen,
+                                         const uint8_t *pk,
+                                         mld_hash_alg_t hashAlg)
+{
+  /* formatted message: 0x01 || ctxlen (1 byte) || ctx || oid || ph */
+  MLD_ALIGN uint8_t fmsg[2 + 255 + 11 + 64];
+  int result;
+
+  if (ctxlen > 255)
+  {
+    return -1;
+  }
+  fmsg[0] = 1;
+  fmsg[1] = ctxlen;
+  if (ctx != NULL && ctxlen != 0)
+  {
+    mld_memcpy(fmsg + 2, ctx, ctxlen);
+  }
+  if (sha_oid(fmsg + 2 + ctxlen, hashAlg, phlen))
+  {
+    return -1;
+  }
+  mld_memcpy(fmsg + 2 + ctxlen + 11, ph, phlen);
```

Comment:
The logic for preparing the formatted message should be hoisted out into a helper which can be used both by sign and verify.

Reply:
done - added `mld_format_pre_hash_message`